### PR TITLE
Allow persistent icons

### PR DESCRIPTION
**Pull request by karimnaaji**
I am not so happy with this solution since it adds a `persistent` keyword in the stylesheet not existing in JS, but that's the easiest workaround now to enable persistent icons. We could detect if the icon comes from a client data source but this would get pretty messy since it introduces a lot of assumptions.

### CODE REVIEW

**Comment by bcamper** [comment]
Thanks @karimnaaji, before we merge let's think a little more on syntax, in case we can think of future uses. For example, maybe something like a `transition` keyword? Default `transition: fade`, with this example being `transition: false`? Not sure if there are other obvious transitions though, so who knows :)  Could also be something like `fade: false` to be more explicit (`persistent` seems a little vague here).

**Comment by karimnaaji** [comment]
Yeah this keyword should be extended to different options to clearly differentiate each cases. 

The fact of colliding or not, the transition with an ease function and the time for the transition.

**Comment by karimnaaji** [comment]
From the current implementation a proposed syntax could also be:

```yaml
icons:
    collide: true | false # defaults to true
    hide:
        effect: fade | blink ...
        ease: pow | sine | ...
        time: 0.2s # sec or millis
    show:
        effect: fade
        ease: pow
        time: 0.1s
```

**Comment by blair1618** [comment]
Seems like we'll need at least one or two more style sheet options for eraser-map functionality:

**1. Enable/disable label fading:** Could be a key like `transition` or `fade`. For semantic density, I'd suggest that the value be the desired duration of the transition in seconds. For our "find me" icon, we'd then use `fade: 0` for an immediate transition. We could allow other values later for more sophisticated behaviors. 

**2. Enable/disable label collision:** A separate key like `collision` would be one approach, but since this behavior is entangled with `priority`, it might make sense to encode collide-ability within that value. For instance, what if we used `priority: 0` to denote that a label always "survives" collision? What this doesn't encode is how particular groups of labels interact - e.g. if we wanted to "mask" collision between certain groups.

**Comment by bcamper** [comment]
```
    transition:
        event: [hide, show]
        effect: fade | blink ...
        ease: pow | sine | ...
        time: 0.2s # sec or millis
```

```
    transition:
        - event: hide
          effect: fade | blink ...
          ease: pow | sine | ...
          time: 0.2s # sec or millis

        - event: show
          effect: fade
          ease: pow
          time: 0.1s
```

**Comment by bcamper** [comment]
```
icons:
    collide: true | false # defaults to true
    hide: &transition
        effect: fade | blink ...
        ease: pow | sine | ...
        time: 0.2s # sec or millis
    show: *transition
```

**Comment by bcamper** [comment]
```
icons:
    collide: true | false # defaults to true
    transition:
        hide:
            effect: fade | blink ...
            ease: pow | sine | ...
            time: 0.2s # sec or millis
        show: hide
        selected: hide
```

**Comment by bcamper** [comment]
```
icons:
    collide: true | false # defaults to true
    [hide, show]:
        effect: fade | blink ...
        ease: pow | sine | ...
        time: 0.2s # sec or millis
    selected:
        effect: fade
        ease: pow
        time: 0.1s
```

**Comment by bcamper** [comment]
```
icons:
    collide: true | false # defaults to true
    transition:
        [hide, show]:
            effect: fade | blink ...
            ease: pow | sine | ...
            time: 0.2s # sec or millis
        selected:
            effect: fade
            ease: pow
            time: 0.1s
```

**Comment by bcamper** [comment]
Minimal implementation:

```
    transition:
        [hide, show]:
            time: 0.2s # sec or millis
        selected:
            time: 0 # turn off effect
```

**Comment by bcamper** [comment]
Initial with `effect: fade | false `

```
    transition:
        [hide, show]:
            effect: fade
            time: 0.2s
        selected:
            effect: false
```

**Comment by karimnaaji** [comment]
This is actually invalid `yaml`:
```yaml
transition:
    [hide, show]:
```
I'll go with
```yaml
transition:
    hide, show:
```

**Comment by blair1618** [comment]
Really? I tried this in a yaml validator just yesterday, to make sure I
wasn't crazy, and it worked fine

**Comment by karimnaaji** [comment]
This one doesn't want it: http://yaml-online-parser.appspot.com/

**Comment by tallytalwar** [comment]
http://www.yamllint.com/ 
it works here.

**Comment by bcamper** [comment]
Yes I believe it is valid YAML, from the spec, but I also think it is not commonly used, so it isn't shocking if some validators/parsers don't support it properly. In JS a `[hide, show]` key just gets de-serialized to `hide,show` anyway (I checked, the brackets are dropped).

```
transition:
    hide, show:
```

is not bad either and is also valid YAML (and to make JS decoding work, we can't allow commas in event names anyway, so there's functionally little or no difference as far as I can tell). But using a true sequence is probably most consistent.

**Comment by blair1618** [comment]
@karimnaaji I just tried the sequence-key syntax in yaml-cpp and it seems to work in a simple test case:
```c++
TEST_CASE("Load a complex key into yaml", "[yaml]") {

    auto str = R"END(
    transitions:
        [show, hide]: 
            effect: fade
            time: 1
        selected:
            effect: blink
            time: 0.5
    )END";

    YAML::Node node = YAML::Load(str);

    REQUIRE(node["transitions"].size() == 2);
}
```
Am I overlooking something?

**Comment by karimnaaji** [comment]
Yep! This works fine, something else in `sceneLoader.cpp` must be assuming something that breaks this and throw an exception then.

File: core/src/scene/sceneLoader.cpp [hunk 1]
```diff
@@ -1134,6 +1133,11 @@ void SceneLoader::parseStyleParams(Node params, Scene& scene, const std::string&
         }
         if (key == "style") { continue; }
 
+        if (key == "transition") {
+            parseTransition(prop.second, scene, key, out);
```
Comment (tallytalwar) [inline core/src/scene/sceneLoader.cpp]: isn't `key` always `transition` here?

File: core/src/scene/sceneLoader.cpp [hunk 2]
```diff
@@ -1249,6 +1253,48 @@ StyleUniforms SceneLoader::parseStyleUniforms(const Node& value, Scene& scene) {
...
+            case YAML::NodeType::Sequence:
+            {
+                auto rawKeys = parseSequence(prop.first);
+                std::stringstream ss(rawKeys);
+                std::vector<std::string> keys;
+                std::string key;
+
+                while (std::getline(ss, key, ',')) {
+                    keys.push_back(key);
+                }
```
Comment (blair1618) [inline core/src/scene/sceneLoader.cpp]: More briefly:
```c++
auto keys = prop.first.as<std::vector<std::string>>();
```
We could also pull `keys` and the loop over `keys` outside the switch/case and just populate`keys` inside the cases, if that makes sense to you. 

Just trying to keep `SceneLoader` on a diet :)

**Comment by blair1618** [comment]
Looking over this now!

**Pull request merged**